File: packages/contracts/src/schemas/credential-delivery.ts
import { Type, type Static } from "../internal.js";
import {
  JSON_SCHEMA_DRAFT_2020_12,
  RUNX_CONTRACT_IDS,
  RUNX_LOGICAL_SCHEMAS,
  type DeepReadonly,
  dateTimeStringSchema,
  generatedSchema,
  stringEnum,
  validateContractSchema,
} from "../internal.js";
import { referenceSchema } from "./spine.js";

const credentialDeliveryModes = ["process_env"] as const;
const credentialDeliveryPurposes = [
  "provider_api",
  "registry",
  "artifact_store",
  "webhook_verification",
] as const;
const credentialMaterialRoles = [
  "personal_token",
  "api_key",
  "client_secret",
  "session_token",
] as const;
const credentialDeliveryStatuses = [
  "delivered",
  "denied",
  "not_found",
  "profile_mismatch",
] as const;
const credentialDeliveryObservationStatuses = [
  "delivered",
  "denied",
  "not_delivered",
] as const;

export const credentialDeliveryModeSchema = stringEnum(credentialDeliveryModes);
export const credentialDeliveryPurposeSchema = stringEnum(credentialDeliveryPurposes);
export const credentialMaterialRoleSchema = stringEnum(credentialMaterialRoles);
export const credentialDeliveryStatusSchema = stringEnum(credentialDeliveryStatuses);
export const credentialDeliveryObservationStatusSchema = stringEnum(
  credentialDeliveryObservationStatuses,
);

export const credentialDeliveryEnvBindingSchema = Type.Object(
  {
    role: credentialMaterialRoleSchema,
    env_var: Type.String({ minLength: 1, pattern: "^[A-Z_][A-Z0-9_]*$" }),
    required: Type.Boolean(),
  },
  { additionalProperties: false },
);

export type CredentialDeliveryEnvBindingContract =
  DeepReadonly<Static<typeof credentialDeliveryEnvBindingSchema>>;

const credentialDeliveryProfileV1TypeSchema = Type.Object(
  {
    schema: Type.Literal(RUNX_LOGICAL_SCHEMAS.credentialDeliveryProfile),
    profile_id: Type.String({ minLength: 1 }),
    provider: Type.String({ minLength: 1 }),
    auth_mode: Type.String({ minLength: 1 }),
    purpose: credentialDeliveryPurposeSchema,
    delivery_mode: credentialDeliveryModeSchema,
    material_roles: Type.Array(credentialMaterialRoleSchema, { minItems: 1 }),
    env_bindings: Type.Array(credentialDeliveryEnvBindingSchema, { minItems: 1 }),
    redaction_policy_ref: referenceSchema,
  },
  {
    $schema: JSON_SCHEMA_DRAFT_2020_12,
    $id: RUNX_CONTRACT_IDS.credentialDeliveryProfile,
    "x-runx-schema": RUNX_LOGICAL_SCHEMAS.credentialDeliveryProfile,
    additionalProperties: false,
  },
);

export type CredentialDeliveryProfileContract =
  DeepReadonly<Static<typeof credentialDeliveryProfileV1TypeSchema>>;

export const credentialDeliveryProfileV1Schema = generatedSchema<CredentialDeliveryProfileContract>(
  "credential-delivery-profile.schema.json",
);

const credentialDeliveryRequestV1TypeSchema = Type.Object(
  {
    schema: Type.Literal(RUNX_LOGICAL_SCHEMAS.credentialDeliveryRequest),
    request_id: Type.String({ minLength: 1 }),
    harness_ref: referenceSchema,
    host_ref: referenceSchema,
    grant_ref: referenceSchema,
    credential_ref: referenceSchema,
    profile_id: Type.String({ minLength: 1 }),
    provider: Type.String({ minLength: 1 }),
    purpose: credentialDeliveryPurposeSchema,
    requested_roles: Type.Array(credentialMaterialRoleSchema, { minItems: 1 }),
    requested_at: dateTimeStringSchema(),
  },
  {
    $schema: JSON_SCHEMA_DRAFT_2020_12,
    $id: RUNX_CONTRACT_IDS.credentialDeliveryRequest,
    "x-runx-schema": RUNX_LOGICAL_SCHEMAS.credentialDeliveryRequest,
    additionalProperties: false,
  },
);

export type CredentialDeliveryRequestContract =
  DeepReadonly<Static<typeof credentialDeliveryRequestV1TypeSchema>>;

export const credentialDeliveryRequestV1Schema = generatedSchema<CredentialDeliveryRequestContract>(
  "credential-delivery-request.schema.json",
);

export const credentialDeliveryHandleSchema = Type.Object(
  {
    role: credentialMaterialRoleSchema,
    delivery_handle_ref: referenceSchema,
    env_var: Type.Optional(Type.String({ minLength: 1, pattern: "^[A-Z_][A-Z0-9_]*$" })),
  },
  { additionalProperties: false },
);

export type CredentialDeliveryHandleContract =
  DeepReadonly<Static<typeof credentialDeliveryHandleSchema>>;

const credentialDeliveryResponseV1TypeSchema = Type.Object(
  {
    schema: Type.Literal(RUNX_LOGICAL_SCHEMAS.credentialDeliveryResponse),
    response_id: Type.String({ minLength: 1 }),
    request_id: Type.String({ minLength: 1 }),
    status: credentialDeliveryStatusSchema,
    delivery_mode: Type.Optional(credentialDeliveryModeSchema),
    handles: Type.Optional(Type.Array(credentialDeliveryHandleSchema)),
    credential_refs: Type.Array(referenceSchema),
    material_ref_hash: Type.Optional(Type.String({ minLength: 1 })),
    denied_reasons: Type.Optional(Type.Array(Type.String({ minLength: 1 }))),
    issued_at: dateTimeStringSchema(),
    expires_at: Type.Optional(dateTimeStringSchema()),
  },
  {
    $schema: JSON_SCHEMA_DRAFT_2020_12,
    $id: RUNX_CONTRACT_IDS.credentialDeliveryResponse,
    "x-runx-schema": RUNX_LOGICAL_SCHEMAS.credentialDeliveryResponse,
    additionalProperties: false,
  },
);

export type CredentialDeliveryResponseContract =
  DeepReadonly<Static<typeof credentialDeliveryResponseV1TypeSchema>>;

export const credentialDeliveryResponseV1Schema =
  generatedSchema<CredentialDeliveryResponseContract>(
    "credential-delivery-response.schema.json",
  );

const credentialDeliveryObservationV1TypeSchema = Type.Object(
  {
    schema: Type.Literal(RUNX_LOGICAL_SCHEMAS.credentialDeliveryObservation),
    observation_id: Type.String({ minLength: 1 }),
    request_id: Type.String({ minLength: 1 }),
    response_id: Type.Optional(Type.String({ minLength: 1 })),
    status: credentialDeliveryObservationStatusSchema,
    harness_ref: referenceSchema,
    host_ref: Type.Optional(referenceSchema),
    profile_id: Type.String({ minLength: 1 }),
    provider: Type.String({ minLength: 1 }),
    purpose: credentialDeliveryPurposeSchema,
    delivery_mode: Type.Optional(credentialDeliveryModeSchema),
    credential_refs: Type.Array(referenceSchema),
    material_ref_hash: Type.Optional(Type.String({ minLength: 1 })),
    delivered_roles: Type.Array(credentialMaterialRoleSchema),
    redaction_refs: Type.Optional(Type.Array(referenceSchema)),
    observed_at: dateTimeStringSchema(),
  },
  {
    $schema: JSON_SCHEMA_DRAFT_2020_12,
    $id: RUNX_CONTRACT_IDS.credentialDeliveryObservation,
    "x-runx-schema": RUNX_LOGICAL_SCHEMAS.credentialDeliveryObservation,
    additionalProperties: false,
  },
);

export type CredentialDeliveryObservationContract =
  DeepReadonly<Static<typeof credentialDeliveryObservationV1TypeSchema>>;

export const credentialDeliveryObservationV1Schema =
  generatedSchema<CredentialDeliveryObservationContract>(
    "credential-delivery-observation.schema.json",
  );

export function validateCredentialDeliveryProfileContract(
  value: unknown,
  label = "credential_delivery_profile",
): CredentialDeliveryProfileContract {
  return validateContractSchema(credentialDeliveryProfileV1Schema, value, label);
}

export function validateCredentialDeliveryRequestContract(
  value: unknown,
  label = "credential_delivery_request",
): CredentialDeliveryRequestContract {
  return validateContractSchema(credentialDeliveryRequestV1Schema, value, label);
}

export function validateCredentialDeliveryResponseContract(
  value: unknown,
  label = "credential_delivery_response",
): CredentialDeliveryResponseContract {
  return validateContractSchema(credentialDeliveryResponseV1Schema, value, label);
}

export function validateCredentialDeliveryObservationContract(
  value: unknown,
  label = "credential_delivery_observation",
): CredentialDeliveryObservationContract {
  return validateContractSchema(credentialDeliveryObservationV1Schema, value, label);
}

File: crates/runx-contracts/src/credential_delivery.rs
//! Credential delivery contracts: public refs, handles, and observations only.
use serde::{Deserialize, Serialize};

use crate::Reference;
use crate::schema::{IsoDateTime, NonEmptyString, RunxSchema};

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
#[serde(rename_all = "snake_case")]
pub enum CredentialDeliveryMode {
    ProcessEnv,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
#[serde(rename_all = "snake_case")]
pub enum CredentialDeliveryPurpose {
    ProviderApi,
    Registry,
    ArtifactStore,
    WebhookVerification,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
#[serde(rename_all = "snake_case")]
pub enum CredentialMaterialRole {
    PersonalToken,
    ApiKey,
    ClientSecret,
    SessionToken,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
#[serde(rename_all = "snake_case")]
pub enum CredentialDeliveryStatus {
    Delivered,
    Denied,
    NotFound,
    ProfileMismatch,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
#[serde(rename_all = "snake_case")]
pub enum CredentialDeliveryObservationStatus {
    Delivered,
    Denied,
    NotDelivered,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
pub enum CredentialDeliveryProfileSchema {
    #[serde(rename = "runx.credential_delivery.profile.v1")]
    V1,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
pub enum CredentialDeliveryRequestSchema {
    #[serde(rename = "runx.credential_delivery.request.v1")]
    V1,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
pub enum CredentialDeliveryResponseSchema {
    #[serde(rename = "runx.credential_delivery.response.v1")]
    V1,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
pub enum CredentialDeliveryObservationSchema {
    #[serde(rename = "runx.credential_delivery.observation.v1")]
    V1,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
#[serde(deny_unknown_fields)]
pub struct CredentialDeliveryEnvBinding {
    pub role: CredentialMaterialRole,
    pub env_var: String,
    pub required: bool,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
#[serde(deny_unknown_fields)]
#[runx_schema(id = "runx.credential_delivery.profile.v1")]
pub struct CredentialDeliveryProfile {
    pub schema: CredentialDeliveryProfileSchema,
    pub profile_id: NonEmptyString,
    pub provider: NonEmptyString,
    pub auth_mode: NonEmptyString,
    pub purpose: CredentialDeliveryPurpose,
    pub delivery_mode: CredentialDeliveryMode,
    pub material_roles: Vec<CredentialMaterialRole>,
    pub env_bindings: Vec<CredentialDeliveryEnvBinding>,
    pub redaction_policy_ref: Reference,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
#[serde(deny_unknown_fields)]
#[runx_schema(id = "runx.credential_delivery.request.v1")]
pub struct CredentialDeliveryRequest {
    pub schema: CredentialDeliveryRequestSchema,
    pub request_id: NonEmptyString,
    pub harness_ref: Reference,
    pub host_ref: Reference,
    pub grant_ref: Reference,
    pub credential_ref: Reference,
    pub profile_id: NonEmptyString,
    pub provider: NonEmptyString,
    pub purpose: CredentialDeliveryPurpose,
    pub requested_roles: Vec<CredentialMaterialRole>,
    pub requested_at: IsoDateTime,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
#[serde(deny_unknown_fields)]
pub struct CredentialDeliveryHandle {
    pub role: CredentialMaterialRole,
    pub delivery_handle_ref: Reference,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub env_var: Option<String>,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
#[serde(deny_unknown_fields)]
#[runx_schema(id = "runx.credential_delivery.response.v1")]
pub struct CredentialDeliveryResponse {
    pub schema: CredentialDeliveryResponseSchema,
    pub response_id: NonEmptyString,
    pub request_id: NonEmptyString,
    pub status: CredentialDeliveryStatus,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub delivery_mode: Option<CredentialDeliveryMode>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub handles: Option<Vec<CredentialDeliveryHandle>>,
    pub credential_refs: Vec<Reference>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub material_ref_hash: Option<NonEmptyString>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub denied_reasons: Option<Vec<NonEmptyString>>,
    pub issued_at: IsoDateTime,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub expires_at: Option<IsoDateTime>,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, RunxSchema)]
#[serde(deny_unknown_fields)]
#[runx_schema(id = "runx.credential_delivery.observation.v1")]
pub struct CredentialDeliveryObservation {
    pub schema: CredentialDeliveryObservationSchema,
    pub observation_id: NonEmptyString,
    pub request_id: NonEmptyString,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub response_id: Option<NonEmptyString>,
    pub status: CredentialDeliveryObservationStatus,
    pub harness_ref: Reference,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub host_ref: Option<Reference>,
    pub profile_id: NonEmptyString,
    pub provider: NonEmptyString,
    pub purpose: CredentialDeliveryPurpose,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub delivery_mode: Option<CredentialDeliveryMode>,
    pub credential_refs: Vec<Reference>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub material_ref_hash: Option<NonEmptyString>,
    pub delivered_roles: Vec<CredentialMaterialRole>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub redaction_refs: Option<Vec<Reference>>,
    pub observed_at: IsoDateTime,
}
